File: app/src/main/java/org/digitalcampus/oppia/adapter/DownloadMediaAdapter.java
package org.digitalcampus.oppia.adapter;


import android.content.Context;
import android.content.SharedPreferences;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ProgressBar;
import android.widget.TextView;

import org.digitalcampus.mobile.learning.R;
import org.digitalcampus.mobile.learning.databinding.RowMediaDownloadBinding;
import org.digitalcampus.oppia.activity.PrefsActivity;
import org.digitalcampus.oppia.listener.ListInnerBtnOnClickListener;
import org.digitalcampus.oppia.model.Course;
import org.digitalcampus.oppia.model.Media;

import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;

import androidx.annotation.NonNull;
import androidx.preference.PreferenceManager;

public class DownloadMediaAdapter extends MultiChoiceRecyclerViewAdapter<DownloadMediaAdapter.DownloadMediaViewHolder> {


    private final SharedPreferences prefs;
    private List<Media> mediaList;
    private final Context context;
    private ListInnerBtnOnClickListener itemClickListener;

    public DownloadMediaAdapter(Context context, List<Media> mediaList) {
        this.context = context;
        this.mediaList = mediaList;
        prefs = PreferenceManager.getDefaultSharedPreferences(context);
    }

    @Override
    public DownloadMediaViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        View contactView = LayoutInflater.from(context).inflate(R.layout.row_media_download, parent, false);
        return new DownloadMediaViewHolder(contactView);
    }


    @Override
    public void onBindViewHolder(@NonNull final DownloadMediaViewHolder viewHolder, final int position) {

        updateViewHolder(viewHolder, position);
        final Media m = getItemAtPosition(position);

        StringBuilder courses =  new StringBuilder();
        for (int i = 0; i < m.getCourses().size(); i++) {
            Course c = m.getCourses().get(i);
            String title = c.getTitle(prefs);
            courses.append( i != 0 ? ", " + title : title);
        }

        viewHolder.binding.mediaCourses.setText(courses.toString());
        viewHolder.binding.mediaTitle.setText(m.getFilename());
        viewHolder.binding.mediaPath.setText(m.getDownloadUrl());
        if (m.getFileSize() != 0) {
            viewHolder.binding.mediaFileSize.setText(context.getString(R.string.media_file_size, m.getFileSize() / (1024 * 1024)));
            viewHolder.binding.mediaFileSize.setVisibility(View.VISIBLE);
        } else {
            viewHolder.binding.mediaFileSize.setVisibility(View.INVISIBLE);
        }

        viewHolder.binding.actionBtn.setVisibility(isMultiChoiceMode ? View.INVISIBLE : View.VISIBLE);

        int actionBtnimagRes;
        if (m.isDownloading()) {
            actionBtnimagRes = R.drawable.ic_action_cancel;

            viewHolder.binding.downloadProgress.setVisibility(View.VISIBLE);
            viewHolder.binding.mediaPath.setVisibility(View.GONE);
            viewHolder.binding.downloadError.setVisibility(View.GONE);
            if (m.getProgress() > 0) {
                viewHolder.binding.downloadProgress.setIndeterminate(false);
                viewHolder.binding.downloadProgress.setProgress(m.getProgress());
            } else {
                viewHolder.binding.downloadProgress.setIndeterminate(true);
            }
        } else {
            actionBtnimagRes = m.hasFailed() ? R.drawable.dialog_ic_action_update : R.drawable.ic_action_download;
            viewHolder.binding.downloadError.setVisibility(m.hasFailed() ? View.VISIBLE : View.GONE);
            viewHolder.binding.downloadProgress.setVisibility(View.GONE);
            viewHolder.binding.mediaPath.setVisibility(View.VISIBLE);
        }
        viewHolder.binding.actionBtn.setImageResource(actionBtnimagRes);
        viewHolder.binding.actionBtn.setTag(actionBtnimagRes);
    }

    @Override
    public int getItemCount() {
        return mediaList.size();
    }

    public Media getItemAtPosition(int position) {
        return mediaList.get(position);
    }

    public void sortByCourse() {
        //Sort the media list by filename
        SharedPreferences prefsCourse = PreferenceManager.getDefaultSharedPreferences(context);
        final String lang = prefsCourse.getString(PrefsActivity.PREF_LANGUAGE, Locale.getDefault().getLanguage());
        Collections.sort(this.mediaList, (Comparator<Object>) (o1, o2) -> {
            String titleCourse1 = ((Media) o1).getCourses().get(0).getTitle(lang);
            String titleCourse2 = ((Media) o2).getCourses().get(0).getTitle(lang);
            return (titleCourse1.compareTo(titleCourse2));
        });

        notifyDataSetChanged();
    }

    public void sortByFilename() {
        //Sort the media list by filename
        Collections.sort(this.mediaList, (Comparator<Object>) (o1, o2) -> ((Media) o1).getFilename().compareTo(((Media) o2).getFilename()));
        notifyDataSetChanged();
    }


    public class DownloadMediaViewHolder extends MultiChoiceRecyclerViewAdapter.ViewHolder {

        private final RowMediaDownloadBinding binding;

        public DownloadMediaViewHolder(View itemView) {

            super(itemView);
            binding = RowMediaDownloadBinding.bind(itemView);

            binding.actionBtn.setOnClickListener(v -> {
                if (itemClickListener != null)
                    itemClickListener.onClick(getAdapterPosition());
            });
        }
    }

    public void setOnItemClickListener(ListInnerBtnOnClickListener listener) {
        this.itemClickListener = listener;
    }
}

File: app/src/main/java/org/digitalcampus/oppia/adapter/DownloadMediaAdapter.kt
package org.digitalcampus.oppia.adapter

import android.content.Context
import android.content.SharedPreferences
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.preference.PreferenceManager
import org.digitalcampus.mobile.learning.R
import org.digitalcampus.mobile.learning.databinding.RowMediaDownloadBinding
import org.digitalcampus.oppia.activity.PrefsActivity
import org.digitalcampus.oppia.adapter.DownloadMediaAdapter.DownloadMediaViewHolder
import org.digitalcampus.oppia.listener.ListInnerBtnOnClickListener
import org.digitalcampus.oppia.model.Media
import java.util.Locale

class DownloadMediaAdapter(private val context: Context, private var mediaList: MutableList<Media>) :
    MultiChoiceRecyclerViewAdapter<DownloadMediaViewHolder>() {

    inner class DownloadMediaViewHolder(itemView: View) : ViewHolder(itemView) {
        val binding: RowMediaDownloadBinding

        init {
            binding = RowMediaDownloadBinding.bind(itemView)
            binding.actionBtn.setOnClickListener {
                itemClickListener?.onClick(getAdapterPosition())
            }
        }
    }

    private val prefs: SharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)
    private var itemClickListener: ListInnerBtnOnClickListener? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DownloadMediaViewHolder {
        val contactView = LayoutInflater.from(context).inflate(R.layout.row_media_download, parent, false)
        return DownloadMediaViewHolder(contactView)
    }

    override fun onBindViewHolder(viewHolder: DownloadMediaViewHolder, position: Int) {
        updateViewHolder(viewHolder, position)
        val m = getItemAtPosition(position)
        val courses = StringBuilder()
        for (i in m.getCourses()!!.indices) {
            val c = m.getCourses()!![i]
            val title = c.getTitle(prefs)
            courses.append(if (i != 0) ", $title" else title)
        }
        viewHolder.binding.mediaCourses.text = courses.toString()
        viewHolder.binding.mediaTitle.text = m.filename
        viewHolder.binding.mediaPath.text = m.downloadUrl
        if (m.fileSize != 0.0) {
            viewHolder.binding.mediaFileSize.text = context.getString(R.string.media_file_size, m.fileSize / (1024 * 1024))
            viewHolder.binding.mediaFileSize.visibility = View.VISIBLE
        } else {
            viewHolder.binding.mediaFileSize.visibility = View.INVISIBLE
        }
        viewHolder.binding.actionBtn.visibility = if (isMultiChoiceMode) View.INVISIBLE else View.VISIBLE
        val actionBtnimagRes: Int
        if (m.isDownloading) {
            actionBtnimagRes = R.drawable.ic_action_cancel
            viewHolder.binding.downloadProgress.visibility = View.VISIBLE
            viewHolder.binding.mediaPath.visibility = View.GONE
            viewHolder.binding.downloadError.visibility = View.GONE
            if (m.progress > 0) {
                viewHolder.binding.downloadProgress.isIndeterminate = false
                viewHolder.binding.downloadProgress.progress = m.progress
            } else {
                viewHolder.binding.downloadProgress.isIndeterminate = true
            }
        } else {
            actionBtnimagRes = if (m.hasFailed()) R.drawable.dialog_ic_action_update else R.drawable.ic_action_download
            viewHolder.binding.downloadError.visibility = if (m.hasFailed()) View.VISIBLE else View.GONE
            viewHolder.binding.downloadProgress.visibility = View.GONE
            viewHolder.binding.mediaPath.visibility = View.VISIBLE
        }
        viewHolder.binding.actionBtn.setImageResource(actionBtnimagRes)
        viewHolder.binding.actionBtn.tag = actionBtnimagRes
    }

    override fun getItemCount(): Int {
        return mediaList.size
    }

    fun getItemAtPosition(position: Int): Media {
        return mediaList[position]
    }

    fun sortByCourse() {
        //Sort the media list by filename
        val prefsCourse = PreferenceManager.getDefaultSharedPreferences(context)
        val lang = prefsCourse.getString(PrefsActivity.PREF_CONTENT_LANGUAGE, Locale.getDefault().language)
        mediaList.sortedWith { o1, o2 ->
            val titleCourse1 = o1.getCourses()!![0].getTitle(lang)
            val titleCourse2 = o2.getCourses()!![0].getTitle(lang)
            titleCourse1.compareTo(titleCourse2)
        }

        notifyDataSetChanged()
    }

    fun sortByFilename() {
        //Sort the media list by filename
        mediaList.sortedBy { it.filename }
        notifyDataSetChanged()
    }

    fun setOnItemClickListener(listener: ListInnerBtnOnClickListener?) {
        itemClickListener = listener
    }
}